### PR TITLE
Refactor and extend HyperbandScheduler

### PR DESCRIPTION
**Pull request by mseeger**
*Issue #, if available:*

*Description of changes:*
Apart from clean-ups (unify common code), this is contributing two new variants of Hyperband:
- Allow jobs in different brackets to compete with each other for promotions
- Allow user to specify rung levels (in case the exponential spacing is not appropriate)

*Testing:*
Apart from unit test, I ran some before/after comparisons. Results are exactly the same for random searchers, and very close for bayesopt searchers (where optimization introduces uncontrollable randomness).

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by szha** [comment]
Job PR-750-1 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-750/1/index.html

**Comment by szha** [comment]
Job PR-750-2 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-750/2/index.html

**Comment by szha** [comment]
Job PR-750-3 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-750/3/index.html

**Comment by mseeger** [comment]
This also has some useful fixes

File: core/src/autogluon/core/scheduler/hyperband.py [hunk 1]
```diff
@@ -341,29 +401,43 @@ def add_job(self, task, **kwargs):
                 'bracket': kwargs['bracket'],
                 'reported_result': None,
                 'keep_case': False}
-            milestones = self.terminator.on_task_add(task, **kwargs)
+            first_milestone = self.terminator.on_task_add(task, **kwargs)[-1]
+        # Register pending evaluation(s) with searcher
+        debug_log = self.searcher.debug_log
```
Comment (zhreshold) [inline core/src/autogluon/core/scheduler/hyperband.py:406]: is debug_log a standard flag in searcher?

Comment (mseeger) [reply core/src/autogluon/core/scheduler/hyperband.py:406]: Yes, it is defined in BaseSearcher, but returns None per default.

**Comment by szha** [comment]
Job PR-750-4 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-750/4/index.html

**Comment by szha** [comment]
Job PR-750-5 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-750/5/index.html

**Comment by mseeger** [comment]
@zhreshold @jwmueller Would be great if you could have a look. This is also fixing an issue with the promotion type HB scheduler, that could lead to errors.

**Review by zhreshold — APPROVED** [review]

File: core/src/autogluon/core/scheduler/hyperband.py [hunk 2]
```diff
@@ -40,25 +46,53 @@
...
 
 
 class HyperbandScheduler(FIFOScheduler):
     r"""Implements different variants of asynchronous Hyperband
 
     See 'type' for the different variants. One implementation detail is when
-    using multiple brackets, task allocation to bracket is done randomly 
-    based on a softmax probability.
+    using multiple brackets, task allocation to bracket is done randomly,
+    based on a distribution inspired by the synchronous Hyperband case.
```
Comment (jwmueller) [inline core/src/autogluon/core/scheduler/hyperband.py:62]: Would be helpful to briefly define "rung", "bracket" in the docs (or at least link to paper where they're defined).

File: core/src/autogluon/core/scheduler/hyperband.py [hunk 3]
```diff
@@ -169,11 +206,22 @@ class HyperbandScheduler(FIFOScheduler):
...
+    rung_system_per_bracket : bool
+        This concerns Hyperband with brackets > 1. When starting a job for a
+        new config, it is assigned a randomly sampled bracket. The larger the
+        bracket, the larger the grace period for the config. If
+        `rung_system_per_bracket` is True, we maintain separate rung level
+        systems for each bracket, so that configs only compete with others
+        started in the same bracket. If False, we use a single rung level
+        system, so that all configs compete with each other. In this case, the
+        bracket of a config only determines the initial grace period, i.e. the
+        first milestone at which it starts competing with others.
```
Comment (jwmueller) [inline core/src/autogluon/core/scheduler/hyperband.py:218]: consider adding sentence clarifying when we might prefer rung_system_per_bracket = True vs False

**Review by jwmueller — APPROVED** [review]
Looks good to me, just had minor comments on clarifying some things to user.

**Comment by szha** [comment]
Job PR-750-7 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-750/7/index.html

**Pull request merged**